crates/starknet-os/src/execution/syscall_handler.rs

```diff
@@ -291,12 +292,56 @@ where
             "No more deployed contracts available to replay".to_string().into_boxed_str(),
         ))?;
 
-        Ok(DeployResponse { contract_address, constructor_retdata })
+        // TODO: hideous clone
+        let mut need_retdata_hack = false;
+        execution_helper.os_input.clone().inspect(|os_input| {
+            let class_hash = os_input
+                .contract_address_to_class_hash
+                .get(&contract_address)
+                .expect("No class_hash for contract_address");
+            let num_constructors =
+                if let Some(compiled_class_hash) = os_input.class_hash_to_compiled_class_hash.get(class_hash) {
+                    let casm = os_input.compiled_classes.get(compiled_class_hash).expect("No CASM");
+                    let num_constructors = casm
+                        .get_cairo_lang_contract_class()
+                        .expect("couldn't get cairo lang class")
+                        .entry_points_by_type
+                        .constructor
+                        .len();
+                    num_constructors
+                } else {
+                    let deprecated_cc = os_input.deprecated_compiled_classes.get(class_hash).expect("no deprecated CC");
+                    let num_constructors = deprecated_cc
+                        .get_starknet_api_contract_class()
+                        .expect("couldn't get starknet api class")
+                        .entry_points_by_type
+                        .get(&starknet_api::deprecated_contract_class::EntryPointType::Constructor)
+                        .expect("should have constructor list")
+                        .len();
+                    num_constructors
+                };
+
+            need_retdata_hack = num_constructors == 0;
+        });
+
+        Ok(DeployResponse { contract_address, constructor_retdata, need_retdata_hack })
     }
 
     fn write_response(response: Self::Response, vm: &mut VirtualMachine, ptr: &mut Relocatable) -> WriteResponseResult {
         write_felt(vm, ptr, response.contract_address)?;
-        write_segment(vm, ptr, response.constructor_retdata)?;
+
+        // Bugfix:
+        // When retdata size is 0, the OS will return retdata as a Int(0) instead of a relocatable
+        // as a result of `cast(0, felt*)`. To work around this, we can write the same here so that
+        // later the OS will assert this value is the same as retdata; that is, both need to be the
+        // same value which is Int(0).
+        if response.need_retdata_hack {
+            log::warn!("Writing Felt::ZERO instead of pointer since retdata size is 0");
+            write_felt(vm, ptr, Felt252::ZERO)?; // casted pointer
+            write_felt(vm, ptr, Felt252::ZERO)?; // size
```

Comment:
@notlesh could you add link to the OS code that returns a NULL pointer?

Reply:
Added link here: [845bcfa](https://github.com/keep-starknet-strange/snos/pull/393/commits/845bcfa7cd5b0a26e0e65721f2e9027bbec2d648)